### PR TITLE
[class-inheritance] Concept Documentation

### PR DESCRIPTION
**Pull request by girijakar**
Concept documentation for issue ~~#2351~~ #2972

### CODE REVIEW

**Comment by BethanyG** [comment]
Hi @girijakar 👋🏽 

Thank you for submitting this PR.  😄   Were you intending to do a PR for issue #2350?  There appears to be a lot of information about class members in this particular about doc that should go with `class-customization` (_the prerequisite exercise to this one_) and not `class-inheritance`.  

The about docs for this concept -`class-inheritance` -should only cover and extend upon the bullet points listed in the **Learning Objectives** section of issue #2351 which are:

>  - more fully understand the uses (_and possible abuses_) of `classes`  in Python
>  - understanding and use of `single inheritance` in Python
>  - understanding and use of `multiple inheritance` in Python
>  - the diamond inheritance problem
>  - Pythons MRO (_method resolution order_)
>  - pros and cos of `mixins` () to isolate `multiple inheritance` issues
>  - when and when not to use `__super__()` to call a parent `class` constructor

Additionally,  the concept introduction doc is missing from the set.  It should be brief introduction to the concept topic.

Are you also going to work on the exercise for this concept?   Both issue #2350 and #2351 are for concept **exercises** as  opposed to concept docs -- although the concept docs can be submitted as part of the concept exercise.

Please let me know -- and if you are only working on the concept docs, I will write up a new issue for you.

**Comment by girijakar** [comment]
Oh!
I want to work on #2351 issue for now. After finishing #2351 I'll try to give #2350 a go.

Can you tell me what other uses of class I can include for the first bullet point?
Is Introduction.md the concept introduction doc?

I apologize for the mistake 😶

**Comment by kotp** [comment]
I have set this to draft as it is not yet ready for review.

> Is Introduction.md the concept introduction doc?

Yes.

> > more fully understand the uses (and possible abuses) of classes in Python
> Can you tell me what other uses of class I can include for the first bullet point?

If you can think of at least 1 or 2, place them in the documentation.  This can grow as you develop the documentation, I am sure you will realize other uses as it gets fleshed out.

* * *
Thank you for working on this!

**Comment by girijakar** [comment]
OK, but introduction.md doc is there in the set.

**Comment by BethanyG** [comment]
@girijakar -  So it is!  _My apologies_ for missing it!  I will give this a more through read-through in the morning my time (_I'm in the pacific time zone_).  will you also be submitting the exercise portion of the issue in this PR or a follow on PR -- or are you only wanting to work on the concept as opposed to the exercise?

**Comment by girijakar** [comment]
@BethanyG I want to work on the concept only for now ☺
Apologies for the delay in reply.

**Comment by BethanyG** [comment]
@girijakar - apologies for the delay!  I've opened a new issue that covers just the concept documents.  It's #2972.  I've linked this PR to it.

File: concepts/class-inheritance/about.md [hunk 1]
```diff
@@ -1,2 +1,291 @@
...
+class Thor(Odin):
+    def func2(self):
+        print("Thor: son of Odin")
+
+
+object = Thor()
+object.func1()
+object.func2()
+```
+The derived class will print both the statements of `func1` and `func2`.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:193]: I think this needs to be a more direct and functional example that uses both attributes and methods, to demonstrate _what_ gets inherited, and _how_ one then adds to or overrides those properties and functionality in the child class.  Recommend taking a look at [this DO article](https://www.digitalocean.com/community/tutorials/understanding-class-inheritance-in-python-3) for some inspiration and examples.   This [real python](https://realpython.com/python3-object-oriented-programming/#inherit-from-other-classes-in-python) article is good as well.

File: concepts/class-inheritance/about.md [hunk 2]
```diff
@@ -1,2 +1,291 @@
...
+
+## Multiple Inheritance
+As we've seen, a class inherits from another class in this scenario. On the other side, multiple inheritance is a feature that allows a class to inherit characteristics and methods from many parent classes.
+```python
+class SubclassName(BaseClass1, BaseClass2, ...):
+    pass
+```
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:200]: I think this section needs to be extended to talk about how multiple inheritance works -- and how the parent classes are read in.  I think it might also be helpful to limit the initial example to only two base classes - one the "core" class, and one the "secondary", to help lead into the diamond problem and then to why mixins are a thing.

As with the single inheritance section above, I think it's important here to show how class attributes, instance attributes, and methods are inherited, and from there, how those get extended or overridden in the child class, so more code examples, and more narrative.

File: concepts/class-inheritance/about.md [hunk 3]
```diff
@@ -1,2 +1,291 @@
...
+    def m(self):
+        print("m of B called")
+class C(A):
+    def m(self):
+        print("m of C called")
+class D(B,C):
+    pass
+```
+If we call an instance x of class D, we will get the output as `m of B called`. But if we interchange the order of inheritance in class D i.e. `Class D(C, D)`. We will get the output as `m of C called`. 
+To solve the diamond problem in python, we will look into a new method `mro()`.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:219]: This isn't a new method, this is how Python solves the diamond problem.  `mro()` exposes what Python is doing, but it doesn't turn it on or off.  Suggest wording closer to what is [here](https://www.educative.io/edpresso/what-is-mro-in-python)

**Review by BethanyG — CHANGES_REQUESTED** [review]
@girijakar -- good first pass on this.  😄   I've left some revision notes, and some notes on sections to expand or remove.  As this is quite a lot of changes, I am going to stop here and let you re-work things before going forward with more review/edits.  Please fee free to reach out to me with any questions or issues you have.

File: concepts/class-inheritance/about.md [hunk 4]
```diff
@@ -1,2 +1,291 @@
...
+|:-- | :-- |
+|Mixin classes tend to be simple because they represent simple orthogonal concepts. | Execution of statements at run time tends to jump around in different mixins, making it hard to follow and debug|
+|Helps us to recombine functionalities with different choices | Potential for long compile times|
+## __super()__
+In a nutshell, `super()` gives us access to methods in a superclass from the subclass that inherits from it.
+`super()` by itself returns a temporary object of the superclass, which may subsequently be used to call the methods of that superclass.
+
+But why we want to use `super()`?
+
+Using `super()` to call already created methods avoids having to rebuild those methods in our subclass and allows us to swap out superclasses with little code modifications.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:290]:
```suggestion
Using `super()` to call already created methods avoids having to rebuild those methods in our subclass and allows us to swap out superclasses with little code modifications.

[four-pillars]: https://www.educative.io/edpresso/what-are-the-four-pillars-of-oops-in-python
```

File: concepts/class-inheritance/about.md [hunk 5]
```diff
@@ -1,2 +1,291 @@
-#TODO: Add about for this concept.
+# About
+
+A `Class` is a template/blueprint representing a group of objects that share common properties and relationships.
```
[the remark this answers is not in the document]
Comment (girijakar) [reply concepts/class-inheritance/about.md:3]: Ok

**Comment by BethanyG** [comment]
@girijakar  - That's awesome news!  Thanks so much for working on this.  I am a bit swamped with work, but will make sure to give this a good look over the weekend.

File: concepts/class-inheritance/about.md (hunk 1 above)
Comment (girijakar) [reply concepts/class-inheritance/about.md:28]: You are right, I was also getting a feeling that the example wasn't upto the mark. I have written a new example.

File: concepts/class-inheritance/about.md [hunk 6]
```diff
@@ -1,2 +1,150 @@
-#TODO: Add about for this concept.
+# About
+
+Inheritance is one of the ['four pillars'][four-pillars] of Object Oriented Programming (`OOP`).
+In situations where only a small amount of functionality needs to be customized for a new class, `inheritance` allows code re-use from one or more parent classes, and can help make programs cleaner and more maintainable.
+
+## Inheritance
+
+`Inheritance` describes `is a kind of` relationship between two or more classes, abstracting common details into super class and storing specific ones in the subclass.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:8]:
```suggestion
`Inheritance` describes `is a kind of` relationship between two or more classes, abstracting common details into super (or _parent_) class and storing specific ones in the subclass (or _child class_).

```

File: concepts/class-inheritance/about.md [hunk 7]
```diff
@@ -1,2 +1,150 @@
...
+## Inheritance
+
+`Inheritance` describes `is a kind of` relationship between two or more classes, abstracting common details into super class and storing specific ones in the subclass.
+To create a child class, specify the parent class name inside the pair of parenthesis, followed by it's name.
+Example
+```python
+class Child(Parent):  
+   pass
+```
+Every child class inherits all the behaviours exhibited by their parent class.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:15]:
```suggestion
Every child class inherits all the behaviors (_attributes, constructors, methods_) exhibited by their parent class.
 
```

File: concepts/class-inheritance/about.md [hunk 8]
```diff
@@ -1,2 +1,150 @@
...
+
+## Single Inheritance
+
+When a derived (or child) class inherits only from one base (or parent) class, it is known as single inheritance.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:19]:
```suggestion
When a derived (or child) class inherits only from one base (or parent) class, it is known as _single inheritance_.

```

File: concepts/class-inheritance/about.md [hunk 9]
```diff
@@ -1,2 +1,150 @@
...
+
+```python
+class Person:
+    def __init__(self, fname, lname):
+        self.fname = fname
+        self.lname = lname
+class Employee(Person):
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:26]:
```suggestion
        self.lname = lname
        
class Employee(Person):

```

File: concepts/class-inheritance/about.md [hunk 10]
```diff
@@ -1,2 +1,150 @@
...
+   pass
+```
+Every child class inherits all the behaviours exhibited by their parent class.
+
+## Single Inheritance
+
+When a derived (or child) class inherits only from one base (or parent) class, it is known as single inheritance.
+
+```python
+class Person:
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:22]:
```suggestion
# The parent or base class.
class Person:

```

File: concepts/class-inheritance/about.md [hunk 11]
```diff
@@ -1,2 +1,150 @@
...
+    def __init__(self, fname, lname):
+        self.fname = fname
+        self.lname = lname
+class Employee(Person):
+    all_employees = []
+    def __init__(self, fname, lname, empid):
+        Person.__init__(self, fname, lname)
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:29]:
```suggestion
        # Using the Parent constructor to create the base object.
        Person.__init__(self, fname, lname)

```

File: concepts/class-inheritance/about.md [hunk 12]
```diff
@@ -1,2 +1,150 @@
...
+        self.empid = empid
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:30]:
```suggestion
        # Adding an attribute specific to the Child class.
        self.empid = empid

```

File: concepts/class-inheritance/about.md [hunk 13]
```diff
@@ -1,2 +1,150 @@
...
+...
+p1 = Person('George', 'smith')
+print(p1.fname, '-', p1.lname)
+e1 = Employee('Jack', 'simmons', 456342)
+e2 = Employee('John', 'williams', 123656)
+print(e1.fname, '-', e1.empid)
+print(e2.fname, '-', e2.empid)
+```
+After running the program we will get the following output
+```cmd
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:45]:
```suggestion
```bash

```

File: concepts/class-inheritance/about.md [hunk 14]
```diff
@@ -1,2 +1,150 @@
...
+        self.lname = lname
+class Employee(Person):
+    all_employees = []
+    def __init__(self, fname, lname, empid):
+        Person.__init__(self, fname, lname)
+        self.empid = empid
+        Employee.all_employees.append(self)
+```
+`Employee` class is derived from `Person`.
+Now, we can create an `Employee` object.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:34]:
```suggestion
Now, we can create an `Employee` object.


```

File: concepts/class-inheritance/about.md [hunk 15]
```diff
@@ -1,2 +1,150 @@
...
+print(e2.fname, '-', e2.empid)
+```
+After running the program we will get the following output
+```cmd
+George - smith
+Jack - 456342
+John - 123656
+```
+## Multiple Inheritance
+As we've seen, a class inherits from another class in this scenario. On the other side, multiple inheritance is a feature that allows a class to inherit characteristics and methods from many parent classes.
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:51]:
```suggestion
As we've seen, `single inheritance` is where a class inherits directly from another class. 
On the other side, `multiple inheritance` is a Python feature that allows a child class to inherit characteristics and methods from more than one parent class.

```

File: concepts/class-inheritance/about.md [hunk 16]
```diff
@@ -1,2 +1,150 @@
...
+```python
+class SubclassName(BaseClass1, BaseClass2, ...):
+    pass
+```
+### Diamond Problem
```
Comment (BethanyG) [inline concepts/class-inheritance/about.md:56]:
```suggestion
### Multiple Inheritance and the Diamond Problem

```

**Comment by BethanyG** [comment]
@girijakar  are you still working on this?  Given its age, I am going to "optimistically merge" this as-is, and if you wish to continue working on revisions, you can submit a follow-on PR.  Just ping and let us know.  

I will file an improvement issue for these docs, and you are more than welcome to then pick that up.  Thanks again for all your work on this!  💙

**Review by BethanyG — APPROVED** [review]
Optimistically merging.
